_posts/2023-01-19-torchwavelets/torchwavelets.Rmd

```diff
@@ -0,0 +1,613 @@
+
+ao %>%
+  model(STL(ao ~ season(period = 33) + season(period = 7) +
+              season(period = 9) + season(period = 174))) %>%
+  components() %>%
+  autoplot()
+
+```
+
+![Decomposition of AO data into trend, seasonal components, and remainder by feasts::STL().](images/stl-nao.png){alt="Decomposition of NAO data into trend, seasonal components, and remainder by feasts::STL()."}
+
+Finally, what can the scaleogram tell us about dominant patterns?
+
+```{r}
+ao_idx <- ao$ao %>% as.numeric() %>% torch_tensor()
+dt <- 1/12 # same interval as for ENSO and NAO
+wtf <- wavelet_transform(length(ao_idx), dt = dt)
+power_spectrum <- wtf$power(ao_idx)
+
+times <- lubridate::year(ao$x) + lubridate::month(ao$x)/12 # also same
+scales <- as.numeric(wtf$scales) # will be same because all series have same length
+
+df <- as_tibble(as.matrix(power_spectrum$t()), .name_repair = "universal") %>%
+  mutate(time = times) %>%
+  pivot_longer(!time, names_to = "scale", values_to = "power") %>%
+  mutate(scale = scales[scale %>%
+    str_remove("[\\.]{3}") %>%
+    as.numeric()])
+
+coi <- wtf$coi(times[1], times[length(ao_idx)])
+coi_df <- data.frame(x = as.numeric(coi[[1]]), y = as.numeric(coi[[2]]))
+
+labeled_scales <- c(0.25, 0.5, 1, 2, 4, 8, 16, 32, 64) # same since scales are same 
+labeled_frequencies <- round(as.numeric(wtf$fourier_period(labeled_scales)), 1)
+
+ggplot(df) +
+  scale_y_continuous(
+    trans = scales::compose_trans(scales::log2_trans(), scales::reverse_trans()),
+    breaks = c(0.25, 0.5, 1, 2, 4, 8, 16, 32, 64),
+    limits = c(max(scales), min(scales)),
+    expand = c(0, 0),
+    sec.axis = dup_axis(
+      labels = scales::label_number(labeled_frequencies),
+      name = "Fourier period (years)"
+    )
+  ) +
+  ylab("scale (years)") +
+  scale_x_continuous(breaks = seq(1950, 2020, by = 5), expand = c(0, 0)) +
+  xlab("year") +
+  geom_contour_filled(aes(time, scale, z = power), show.legend = FALSE) +
+  scale_fill_viridis_d(option = "turbo") +
+  geom_ribbon(data = coi_df, aes(x = x, ymin = y, ymax = max(scales)),
+              fill = "black", alpha = 0.6) +
+  theme(legend.position = "none")
+
+```
+
+![Scaleogram of AO data.](images/scaleogram-ao.png)
+
+Having seen the overall spectrum, the lack of strongly dominant patterns in the scaleogram does not come as a big surprise. It is tempting -- for me, at least -- to see a reflection of ENSO around 1970, all the more since by transitivity, AO and ENSO should be related in some way. But here, qualified judgment really is reserved to the experts.
+
+## Conclusion
+
+Like I said in the beginning, this post would be about inspiration, not technical detail or reportable results. And I hope that inspirational it has been, at least a little bit. If you're experimenting with wavelets yourself, or plan to -- or if you work in the atmospheric sciences, and would like to provide some insight on the above data/phenomena -- we'd love to hear from you!
```

Comment:
```suggestion
Like I said in the beginning, this post is about inspiration, not technical detail or reportable results. And I hope that inspirational it has been, at least a little bit. If you're experimenting with wavelets yourself, or plan to -- or if you work in the atmospheric sciences, and would like to provide some insight on the above data/phenomena -- we'd love to hear from you!
```

Reply:
Slightly more straightforward way of saying it.

Reply:
yeah, I get it ... sorry but I'll skip this one, I think I need the ... well ... you see what I mean